### PR TITLE
III-3585 Add Sentry integration in amqp consumers

### PR DESCRIPTION
**Pull request by bertramakers**
### Added
 
- Added Sentry integration inside AMQP consumer.
 
---

Ticket: https://jira.uitdatabank.be/browse/III-3585

**Note**: I had to move some stuff around to make this work. Best to review commit per commit.

**Context**:
The `AbstractConsumer` which is used by most AMQP consumers in our apps, catches all exceptions and logs them using a PSR logger: https://github.com/cultuurnet/broadway-amqp/blob/master/src/AbstractConsumer.php#L158-L179

So instead of making a change to that library to also make it possible to inject a Sentry class there, I've created a decorator for PSR loggers that also sends the errors and emergencies to Sentry.

Tested it by temporarily adding a `throw new Exception()` inside `EventSearchProjector` which handles `EventProjectedToJSONLD` messages coming from AMQP.

It still showed up in the logs:
```
[2020-10-23 15:40:57] amqp.udb3_publisher.INFO: passing on message to event bus {"correlation_id":"1dabb930-297c-4f4d-98ff-123e6ad75606-1"} []
[2020-10-23 15:40:57] amqp.udb3_publisher.ERROR: Something went wrong while indexing! {"correlation_id":"1dabb930-297c-4f4d-98ff-123e6ad75606-1","exception":"[object] (Exception(code: 0): Something went wrong while indexing! at /var/www/udb-search/src/Event/EventSearchProjector.php:27)"} []
[2020-10-23 15:40:57] amqp.udb3_publisher.INFO: message rejected {"correlation_id":"1dabb930-297c-4f4d-98ff-123e6ad75606-1"} []
```

And also on Sentry (in a development project so as not to pollute Slack): https://sentry.io/organizations/publiq-vzw/issues/1975957720/?project=5480138&query=is%3Aunresolved

The reason I had to move some stuff around was that I wanted to use the `SentryExceptionHandler` class that @LucWollants wrote which includes tags for the API key and more importantly, the environment ("cli" or "web"). But it was instantiated "on the fly" for the Whoops error handlers. So I had to move it to two service providers, one for the CLI and one for web (because the instantiation differs).

Here's a Sentry error that proves that the API key is still being included for web errors: https://sentry.io/organizations/publiq-vzw/issues/1975980338/?project=5480138&query=is%3Aunresolved ("test_api_key")

Both Sentry errors also still have the right `runtime.env` tag (`cli` or `web`)

### CODE REVIEW

File: app/Error/SentryPsrLoggerDecorator.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    private $logger;
+
+    public function __construct(SentryExceptionHandler $sentryHandler, LoggerInterface $logger)
+    {
+        $this->sentryHandler = $sentryHandler;
+        $this->logger = $logger;
+    }
+
+    public function error($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:35]:
```suggestion
    public function error($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->sendThrowableToSentry($context);
+        $this->logger->error($message, $context);
+    }
+
+    public function emergency($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:41]:
```suggestion
    public function emergency($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->sendThrowableToSentry($context);
+        $this->logger->emergency($message, $context);
+    }
+
+    public function alert($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:47]:
```suggestion
    public function alert($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->alert($message, $context);
+    }
+
+    public function critical($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:52]:
```suggestion
    public function critical($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->critical($message, $context);
+    }
+
+    public function warning($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:57]:
```suggestion
    public function warning($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->warning($message, $context);
+    }
+
+    public function notice($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:62]:
```suggestion
    public function notice($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->notice($message, $context);
+    }
+
+    public function info($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:67]:
```suggestion
    public function info($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->info($message, $context);
+    }
+
+    public function debug($message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:72]:
```suggestion
    public function debug($message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->debug($message, $context);
+    }
+
+    public function log($level, $message, array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:77]:
```suggestion
    public function log($level, $message, array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $this->logger->log($level, $message, $context);
+    }
+
+    private function sendThrowableToSentry(array $context = array()): void
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:82]:
```suggestion
    private function sendThrowableToSentry(array $context = []): void
```

File: app/Error/SentryPsrLoggerDecorator.php [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $throwable = $context['exception'] ?? null;
+        if (!$throwable || !($throwable instanceof Throwable)) {
+            return;
+        }
+
+        $inspector = new Inspector($throwable);
+        $this->sentryHandler->setInspector($inspector);
+        $this->sentryHandler->handle();
+    }
```
Comment (jdrieghe) [inline app/Error/SentryPsrLoggerDecorator.php:92]:
```suggestion
        if ($throwable instanceof Throwable) {
            $inspector = new Inspector($throwable);
            $this->sentryHandler->setInspector($inspector);
            $this->sentryHandler->handle();
        }
    }
```

**Review by jdrieghe — DISMISSED** [review]

**Comment by bertramakers** [comment]
@jdrieghe you're right we should use the square bracket syntax. It was auto-generated like this by PHPStorm and I overlooked it. I didn't notice the "Add suggestion to batch" button at first so I did a search & replace locally and commited that instead and then resolved the conversation. Next time I'll just batch the actual suggestions into one commit 👍

**Review by LucWollants — APPROVED** [review]

**Pull request merged**